goodgodgd/vode-2020
# Applying gradient tape

- move evaluation codes to `evalutate/eval_funcs.py`
- create functions to evaluate metrics from batch tensor and test functions for them
    - calc_trajectory_error_tensor()
    - calc_rotational_error_tensor()
    - test_calc_trajectory_error_tensor()
    - test_calc_rotational_error_tensor()
- build only prediction model, training loss and metric are computed after inference
    - create_model()
    - compute_loss_vode()
    - compute_metric_pose()
- add `length` key to `tfr_config.txt` to quickly read the total number of samples in tfrecords
- **apply gradient tape** in training and validation in `model/model_main.py`
    - now training process is controlled manually
- **apply tf.function** in training and validation in `model/model_main.py`
    - train_a_batch(), validate_a_batch()
    - revised some functions to run them in graph mode
    - **graph mode is twice as fast as eager mode**
- leave eager mode training and validation functions as **_eager()
- save log into a file `history.txt` besides check point files
    - save_log()
    - save mean loss and metrics every epoch
    - columns: [epoch,train_loss,train_metric_trj,train_metric_rot,val_loss,val_metric_trj,val_metric_rot]
- read initial epoch from `history.txt`
    - read_previous_epoch()
